### PR TITLE
utmp-wtmp: don't crash if user is not set

### PR DESCRIPTION
**Pull request by dtardon**

### CODE REVIEW

File: src/shared/utmp-wtmp.c [hunk 1]
```diff
@@ -201,7 +201,7 @@ int utmp_put_init_process(const char *id, pid_t pid, pid_t sid, const char *line
 
         if (ut_type == USER_PROCESS) {
                 store.ut_type = USER_PROCESS;
-                strncpy(store.ut_user, user, sizeof(store.ut_user)-1);
+                strncpy(store.ut_user, user ?: "", sizeof(store.ut_user)-1);
```
Comment (bluca) [inline src/shared/utmp-wtmp.c]: strempty(user)

Comment (poettering) [inline src/shared/utmp-wtmp.c]: strempty()

Comment (bluca) [reply src/shared/utmp-wtmp.c]: copycat

Comment (bluca) [reply src/shared/utmp-wtmp.c:204]: copycat

**Comment by poettering** [comment]
so, hmm, this would mean we'd write an invalid utmp entry. I think we should rather fill in the invoking user's name, i.e. this would result int "root", which is the right approach. i.e. in exec_invoke() when utmp is enabled and `username` variable is null, invoke getusername_malloc() locally and pass that along. 

and then add an assert() to utmp_put_init_process() on the username param.

**Comment by YHNdnzj** [comment]
> so, hmm, this would mean we'd write an invalid utmp entry. I think we should rather fill in the invoking user's name, i.e. this would result int "root", which is the right approach. i.e. in exec_invoke() when utmp is enabled and `username` variable is null, invoke getusername_malloc() locally and pass that along.
> 
> and then add an assert() to utmp_put_init_process() on the username param.

Instead, maybe refuse/ignore `UtmpMode=user` if `User=` is not specified? Currently this just crashes, so won't break backwards compat I guess...

**Comment by yuwata** [comment]
-> #31384.

**Comment by yuwata** [comment]
> > so, hmm, this would mean we'd write an invalid utmp entry. I think we should rather fill in the invoking user's name, i.e. this would result int "root", which is the right approach. i.e. in exec_invoke() when utmp is enabled and `username` variable is null, invoke getusername_malloc() locally and pass that along.
> > and then add an assert() to utmp_put_init_process() on the username param.
> 
> Instead, maybe refuse/ignore `UtmpMode=user` if `User=` is not specified? Currently this just crashes, so won't break backwards compat I guess...

I think there is no reason we should refuse or ignore `UtmpMode=user` without `User=` setting.

**Pull request closed without merging**